src/resolvelib/resolvers.py

```diff
@@ -99,11 +95,11 @@ def merged_with(self, provider, requirement, parent):
         """Build a new instance from this and a new requirement."""
         infos = list(self.information)
         infos.append(RequirementInformation(requirement, parent))
-        candidates = provider.find_matches([r for r, _ in infos])
-        if not isinstance(candidates, collections_abc.Sequence):
-            candidates = list(candidates)
-        criterion = type(self)(candidates, infos, list(self.incompatibilities))
-        if not candidates:
+        cands = Candidates(provider.find_matches([r for r, _ in infos]))
+        criterion = type(self)(cands, infos, list(self.incompatibilities))
+        try:
+            next(iter(cands))
+        except StopIteration:
```

Comment:
I suggest defining a `isempty` function to make this (and the one below) looks less complex.